node/src/manager/commands/test_run.rs

```diff
@@ -0,0 +1,577 @@
+
+    let blockchain_map = Arc::new(blockchain_map);
+    let subgraph_instance_manager = SubgraphInstanceManager::new(
+        &logger_factory,
+        subgraph_store.clone(),
+        blockchain_map.clone(),
+        metrics_registry.clone(),
+        link_resolver.cheap_clone(),
+    );
+
+    // Create IPFS-based subgraph provider
+    let subgraph_provider = Arc::new(IpfsSubgraphAssignmentProvider::new(
+        &logger_factory,
+        link_resolver.cheap_clone(),
+        subgraph_instance_manager,
+    ));
+
+    let panicking_subscription_manager = Arc::new(PanicSubscriptionManager {});
+
+    let subgraph_registrar = Arc::new(IpfsSubgraphRegistrar::new(
+        &logger_factory,
+        link_resolver.cheap_clone(),
+        subgraph_provider.clone(),
+        subgraph_store.clone(),
+        panicking_subscription_manager,
+        blockchain_map,
+        node_id.clone(),
+        SubgraphVersionSwitchingMode::Instant,
+    ));
+
+    let (name, hash) = if subgraph.contains(':') {
+        let mut split = subgraph.split(':');
+        (split.next().unwrap(), split.next().unwrap().to_owned())
+    } else {
+        ("cli", subgraph)
+    };
+
+    let subgraph_name = SubgraphName::new(name)
+        .expect("Subgraph name must contain only a-z, A-Z, 0-9, '-' and '_'");
+    let subgraph_hash = DeploymentHash::new(hash).expect("Subgraph hash must be a valid IPFS hash");
+
+    info!(&logger, "Creating subgraph {}", name);
+    let create_result =
+        SubgraphRegistrar::create_subgraph(subgraph_registrar.as_ref(), subgraph_name.clone())
+            .await?;
+
+    info!(
+        &logger,
+        "Looking up subgraph deployment {} (Deployment hash => {}, id => {})",
+        name,
+        subgraph_hash,
+        create_result.id,
+    );
+
+    SubgraphRegistrar::create_subgraph_version(
+        subgraph_registrar.as_ref(),
+        subgraph_name.clone(),
+        subgraph_hash.clone(),
+        node_id.clone(),
+    )
+    .await?;
+
+    // let deployment_locator = DeploymentLocator::new(DeploymentId(deployment_id), subgraph_hash);
+
+    let deployments = Deployment::lookup(&primary, name.to_string())?;
+    let deployment = deployments
+        .first()
+        .expect("At least one deployment should exist");
+
+    SubgraphAssignmentProvider::start(subgraph_provider.as_ref(), deployment.locator(), Some(stop_block)).await?;
+
+    loop {
+        tokio::time::sleep(Duration::from_millis(1000)).await;
+
+        let block_ptr = subgraph_store
+            .least_block_ptr(&subgraph_hash)
+            .unwrap()
+            .unwrap();
+
+        if block_ptr.number >= stop_block {
+            info!(&logger, "subgraph now at block {}, reached stop block {}", block_ptr.number, stop_block);
+            break;
+        }
+
+    }
+    info!(&logger, "Removing subgraph {}", name);
+    subgraph_store.clone().remove_subgraph(subgraph_name)?;
+
+    Ok(())
+}
+
+// Stuff copied directly moslty from `main.rs`
```

Comment:
Everything below that point has been copied from `main.rs`, we will re-use that in some shared code.